chia/wallet/puzzles/load_clvm.py

```diff
@@ -1,53 +1,106 @@
+if "CLVM_TOOLS" in os.environ:
     try:
+        from clvm_tools.clvmc import compile_clvm as compile_clvm_py_candidate
 
+        compile_clvm_py = compile_clvm_py_candidate
+    finally:
+        pass
+
+
+def compile_clvm_in_lock(full_path, output, search_paths):
+    # Compile using rust (default)
+
+    # Ensure path translation is done in the idiomatic way currently
+    # expected.  It can use either a filesystem path or name a python
+    # module.
+    treated_include_paths = list(map(translate_path, search_paths))
+    res = compile_clvm_rust(str(full_path), str(output), treated_include_paths)
+
+    if "CLVM_TOOLS" in os.environ and os.environ["CLVM_TOOLS"] == "check" and compile_clvm_py is not None:
+        # Simple helper to read the compiled output
         def sha256file(f):
             import hashlib
 
             m = hashlib.sha256()
-            m.update(open(f).read().encode("utf8"))
+            m.update(open(f).read().strip().encode("utf8"))
             return m.hexdigest()
 
-        from clvm_tools_rs import compile_clvm as compile_clvm_rs
+        orig = "%s.orig" % output
+
+        compile_clvm_py(full_path, orig, search_paths=search_paths)
+        orig256 = sha256file(orig)
+        rs256 = sha256file(output)
+
+        if orig256 != rs256:
+            print("Compiled %s: %s vs %s\n" % (full_path, orig256, rs256))
+            print("Aborting compilation due to mismatch with rust")
+            assert orig256 == rs256
+        else:
+            print("Compilation match %s: %s\n" % (full_path, orig256))
 
-        def translate_path(p_):
-            p = str(p_)
-            if os.path.isdir(p):
-                return p
-            else:
-                module_object = importlib.import_module(p)
-                return os.path.dirname(inspect.getfile(module_object))
+    return res
 
-        def rust_compile_clvm(full_path, output, search_paths=[]):
-            treated_include_paths = list(map(translate_path, search_paths))
-            compile_clvm_rs(str(full_path), str(output), treated_include_paths)
 
-            if os.environ["CLVM_TOOLS_RS"] == "check":
-                orig = str(output) + ".orig"
-                compile_clvm_py(full_path, orig, search_paths=search_paths)
-                orig256 = sha256file(orig)
-                rs256 = sha256file(output)
+def compile_clvm(full_path, output, search_paths=[]):
+    # Ensure that we're the only one recompiling full_path
+    lock_filename = f"{full_path}.lock"
 
-                if orig256 != rs256:
-                    print("Compiled %s: %s vs %s\n" % (full_path, orig256, rs256))
-                    print("Aborting compilation due to mismatch with rust")
-                    assert orig256 == rs256
+    lock_file = None
+    while True:
+        lock_file = create_exclusive_lock(lock_filename)
+        if lock_file is not None:
+            break
 
-        compile_clvm = rust_compile_clvm
+        time.sleep(0.1)
```

Comment:
perhaps not for this PR. but this looks like a perfect candidate for a context manager, that loops trying to acquire the lock, and then releases it when leaving the context manager.

Reply:
Makes sense.  It was a somewhat desperate move as I didn't want to leave everyone waiting while a concurrency problem caused grief.  Amine helped me come to the decision to do a heavy handed solution but I think I can probably eliminate it with more open time for investigation.

Reply:
I agree not to do it in this PR